### PR TITLE
add lgbm pruning example w/ custom cv fun

### PR DESCRIPTION
**Pull request by mirekphd**
## Motivation
As a follow-up to #1852 (solved with the kind help of @toshihikoyanase ), and possibly contributing also some ideas towards solving #1119 (as noted by @bs1119 in #1852) , this PR is aimed to demonstrate how Optuna pruners can be combined with **custom cross-validation functions**, both for classification and regression models, with early stopping and without. 

In this example, we use Optuna to optimize a classification or regression model, with a custom CV function `custom_cv_fun()` applying LightGBM's `.fit()` method to each CV fold separately, instead of using standard `lgbm.cv()`. This modification aims to improve out-of-fold (CV) model accuracy, by using and individual, data-derived number of iterations for each fold model, which in theory should reduce under- and overfitting, thus improving accuracy of the standard `lightgbm.cv` implementation that uses a single common number of iterations for each fold (even when early stopping is enabled). Deterioration in computational performance is traded-off for improved predictive performance.

Throughout training, an Optuna pruner observes model metrics returned by the custom CV function and - if necessary - stops early unpromising trials via a callback, resulting in noticeably faster execution times with unaffected accuracy.

Special attention is paid to model reproducibility, with precise control over all relevant seed values, to let users arrive at the same results over repeated studies. Execution time comparisons between pruning and no-pruning setups required adding controls for the numbers of threads and workers used by both Optuna and the underlying `lightgbm` methods (but subject to reproducibility constraints which unfortunately reduce Optuna's multiprocessing performance).

## Description of the changes
A single independent python 3 script conforming to existing conventions is added to the folder dedicated for pruning examples.
The script was developed and tested under `lightgbm==3.0.0` with `optuna==2.1.0` in python 3.8, and reformatted with `black`.

### CODE REVIEW

**Comment by mirekphd** [comment]
The funny thing is that the formatting "problems" that `black` is now finding in my file causing "Checks / check" to fail (see [Details](https://github.com/optuna/optuna/pull/1896/checks?check_run_id=1206183006)) were introduced... by `black` itself...:) with an apparently narrower maximum line width on my side than on yours. I can undo those black's style changes manually on my side to match your version, but they will re-emerge next time I use it, so we had better find the reason behind them and file it as an Issue in `black` to prevent similar future problems.

**Comment by HideakiImamura** [comment]
What is the version of `black` in your local environment? The latest `black` is `20.8b1` and you have to the latest version.

**Comment by mirekphd** [comment]
> 
> 
> What is the version of `black` in your local environment? The latest `black` is `20.8b1` and you have to the latest version.

Looks like we have the same version:
```
$ black --version
black, version 20.8b1

$ pip freeze | grep black
black==20.8b1
```

I wander: is black perhaps unstable over multiple runs...? I did run it definitely more than once on this file and initially these problematic lines looked OK - just like "your" black wants them to look...

**Comment by HideakiImamura** [comment]
> I wonder: is black perhaps unstable over multiple runs...?

Hmm... I don't think so. I usually run `black` multiple times in my local environment, but I have no bad effects. 

Do you run `black .` at the optuna directory (the cloned directory from https://github.com/optuna/optuna)? In addition, you have to use Optuna's [pyproject.toml](https://github.com/optuna/optuna/blob/master/pyproject.toml), which includes the setting of the `black`. (It is contained in the optuna directory as a default if you cloned https://github.com/optuna/optuna.)

**Comment by toshihikoyanase** [comment]
@mirekphd Maybe this is not related to your issue, but I encountered a similar issue when I use `black` extension of vscode. I couldn't fix the issue and I stop using the extension now. If you use IDEs, could you run `black` in your terminal?

(Please ignore my comment, if you don't use IDE.)

**Comment by mirekphd** [comment]
I believe `black`'s default max line length is 88 characters (see [docs](https://test-black.readthedocs.io/en/latest/the_black_code_style.html#line-length)), and you bumped it to a more permissive 99 chars here:
https://github.com/optuna/optuna/blob/3163606048374becbec7f238026e8d18c9ef8310/pyproject.toml#L2
which would explain nicely why my "blackened" code looks narrower than that expected by CICD's `black`.

Later in the evening I will revert the code to pre-`black` version and re-run `black` with `--line-length 99` to see if I can replicate your results.

@toshihikoyanase : you can perhaps try the same method if the VSC extension can be parametrized in such a way.

**Comment by mirekphd** [comment]
> I believe `black`'s default max line length is 88 characters (see [docs](https://test-black.readthedocs.io/en/latest/the_black_code_style.html#line-length)), and you bumped it to a more permissive 99 chars here:
> 
> https://github.com/optuna/optuna/blob/3163606048374becbec7f238026e8d18c9ef8310/pyproject.toml#L2
> 
> which would explain nicely why my "blackened" code looks narrower than that expected by CICD's `black`.

Yes, different `black` line length settings are the reason behind the problem with the `Checks/check black` test failure we have above.

As a proof, compare output of those two "black diffs":

`optuna/optuna` settings:
```
$ black --diff --color --line-length 99 lightgbm_with_custom_cv_function.py
```

vs. `black`'s default:
```
$ black --diff --color --line-length 88 lightgbm_with_custom_cv_function.py
```

applied to the original file: 
[lightgbm_with_custom_cv_function.zip](https://github.com/optuna/optuna/files/5354607/lightgbm_with_custom_cv_function.zip)

IMHO we could easily prevent this confusion from repeating itself for other long-line coders in the future by:
- documenting (e.g. [here](https://github.com/optuna/optuna/blob/master/CONTRIBUTING.md)) the non-standard line length, 
- passing the custom `--line-length=99` explicitly as an argument of `black` in the CICD command line, because it currently has no arguments, which suggests that `black`'s defaults can be used to reproduce the expected formatting.

**Comment by mirekphd** [comment]
`flake8` has `--max-line-length` defaulting to 99, but nevertheless it raised 4 "line too long" errors even after passing `black --line-length 99` check, because unlike `black`, `flake8` checks also line length within the docstring...:
```
./examples/pruning/lightgbm_with_custom_cv_function.py:4:100: E501 line too long (154 > 99 characters)
./examples/pruning/lightgbm_with_custom_cv_function.py:6:100: E501 line too long (337 > 99 characters)
./examples/pruning/lightgbm_with_custom_cv_function.py:8:100: E501 line too long (206 > 99 characters)
./examples/pruning/lightgbm_with_custom_cv_function.py:10:100: E501 line too long (173 > 99 characters)
```

**Comment by mirekphd** [comment]
Now that automated checks have been finally passed, I suppose we are ready for the proper code review?:)

**Comment by toshihikoyanase** [comment]
Sorry for the delayed response. I think the code will help LGBM users, and I'll review the code in a day or two.

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
+# Functions
+
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:2]: I think this comment can be removed.
```suggestion
```

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+    # get feature names
+    feature_names = list(X.columns)
+
+    # split data into k folds
+    kf = KFold(n_splits=nfold, shuffle=True, random_state=random_state)
+    splits = kf.split(X, y)
+
+    # iterate over folds
+    for n, (train_index, valid_index) in enumerate(splits):
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:72]: `n` is not used and we can remove it.
```suggestion
    for train_index, valid_index in splits:
```

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+        seed=SEED,
+        verbosity=VERBOSE_LGBM,
+    )
+
+    study = optuna.create_study(sampler=OPTUNA_SAMPLER, pruner=OPTUNA_PRUNER, direction=OPT_DIR)
+
+    start_time = datetime.now()
+
+    # run Optuna optimization over the specified number of trials
+    study.optimize(objective, n_trials=TRIALS, n_jobs=WORKERS_NUM)
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:369]: IMO, `n_jobs>1` will hardly improve the performance due to the GIL. How about stop using it?
Even so, `n_jobs` of LightGBM will utilize the multi-core processors effectively.

```suggestion
    study.optimize(objective, n_trials=TRIALS)
```

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        "seed": SEED,
+        "verbosity": VERBOSE_LGBM,
+    }
+    all_best_params = {**optimized_best_params, **static_params}
+    optuna_best_params["custom"] = all_best_params
+
+    print(
+        "\nBest %s metric for optuna+custom fun. (CV models mean reported by Optuna): %.5f\n"
+        % (METRIC, optuna_best_metrics["custom"])
+    )
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:397]: Please use `format` instead of `%`.
```suggestion
    print(
        "\nBest {} metric for optuna+custom fun. (CV models mean reported by Optuna): {:.5f}\n".format(
            METRIC, optuna_best_metrics["custom"]
        )
    )
```

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+
+        # collect current fold data
+        fold_best_iterations.append(model.best_iteration_)
+        fold_best_scores.append(model.best_score_[eval_name])
+
+    # average folds iterations numbers
+    folds_data = {}
+    folds_data["best_iterations_mean"] = -1
+    if fold_best_iterations[0] is not None:
+        folds_data["best_iterations_mean"] = int(np.mean(fold_best_iterations))
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:111]: [Question] Could you tell me when the `model.best_iteration_` returns `None`?

File: examples/pruning/lightgbm_with_custom_cv_function.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+        valid_y_df = pd.DataFrame({"y": valid_y})
+
+        return train_x_df, valid_x_df, train_y_df, valid_y_df
+
+    else:
+        return train_x, valid_x, train_y, valid_y
+
+
+class Objective_custom:
```
Comment (toshihikoyanase) [inline examples/pruning/lightgbm_with_custom_cv_function.py:167]: Please use camel case (c.f., [coding style convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#example))
```suggestion
class ObjectiveCustom:
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your PR. This example has many code comments, and therefore it is quite kind for readers. On the other hand, it is a bit longer than the other examples. If we omit some unnecessary parts, I think users can focus more on the core part of this example (i.e., `cusotm_cv_fun` and `Objective_custom`). What do you think?

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 3 above)
Comment (mirekphd) [reply examples/pruning/lightgbm_with_custom_cv_function.py:369]: Here I have to kindly disagree:) From my experience on multi-CPU machines (in docker containers), muli-threading built into LGBM has its limits (notably, by Guolin Ke's own admission in one discussion on Githhub) LGBM internal parallelism does not work beyond certain point (specifically, when the user tries to access threads from more than 1 physical CPU socket). In practice, the optimum is around 8-16 threads, depending on data shape. You can see some of my public results trying to establish such optimum [here](https://github.com/mirekphd/gbdt-algos-performance-tests/blob/master/20200812/analyze-results-of-gpu-vs-cpu-tests-multiple-versions-20200812.ipynb).

Enter multi-processing (tested either with `loky` or `multiprocessing` backend, as `dask` does not work), which allows us to bypass this limitation of multithreading built into LightGBM, because processes can be run on separate CPU sockets. If you wish, I can supply results of additional performance tests of this PR script, e.g. time taken on average by an Optuna trial) for the multiprocessing vs. multithreading case?

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 5 above)
Comment (mirekphd) [reply examples/pruning/lightgbm_with_custom_cv_function.py:111]: Apologies - here the comment would be probably in order. This occurs when the user switches off early stopping (by setting `USE_STOPPING=False`, and then the `fold_best_iterations` list is populated with `None` elements (as returned by `model.best_iteration_` in such cases).

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 6 above)
Comment (mirekphd) [reply examples/pruning/lightgbm_with_custom_cv_function.py:167]: For consistency this change is also required in [#L349](https://github.com/mirekphd/optuna/blob/lgbm-w-cust-cv-fun/examples/pruning/lightgbm_with_custom_cv_function.py#L349)

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 3 above)
Comment (mirekphd) [reply examples/pruning/lightgbm_with_custom_cv_function.py:369]: The most likely reason why multiprocessing did not seem to work "out-of-the-box" was this reproducibility enforcement in my example attained by... switching off multiprocessing by setting `WORKERS_NUM` back from 2 to 1 (in line [#L311](https://github.com/mirekphd/optuna/blob/lgbm-w-cust-cv-fun/examples/pruning/lightgbm_with_custom_cv_function.py#L311)).

We might change the current setting of `MAKE_REPRODUCIBLE` from `True` to `False` or set `WORKERS_NUM` to 1 instead of 2 (to require action that makes it more likely to notice the role of `MAKE_REPRODUCIBLE`) if you think others will also make the same conclusions about GIL being the culprit here, rather than the desire to obtain the same result at each Optuna run.

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 4 above)
Comment (mirekphd) [reply examples/pruning/lightgbm_with_custom_cv_function.py:397]: I replaced `%`-style formatting with `{}`-style formatting also in [#L373](https://github.com/mirekphd/optuna/blob/lgbm-w-cust-cv-fun/examples/pruning/lightgbm_with_custom_cv_function.py#L373).

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 3 above)
Comment (toshihikoyanase) [reply examples/pruning/lightgbm_with_custom_cv_function.py:370]: Thank you for your detailed comments. 

> `MAKE_REPRODUCIBLE `

I agree. To make the optimization reproducible, we need to set `n_jobs=1`. So, I think it's reasonable to set `WORKERS_NUM` to `1` by default.

>  switching off multiprocessing by setting `WORKERS_NUM` back from 2 to 1

A nit, but Optuna utilizes the multi-threading, not multiprocessing when we set `n_jobs>1`.

File: examples/pruning/lightgbm_with_custom_cv_function.py (hunk 5 above)
Comment (toshihikoyanase) [reply examples/pruning/lightgbm_with_custom_cv_function.py:110]: Thank you for adding the comment to L107 :+1:

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update.

I'm still wondering if we can make this example short. The lengths of the other examples are less than 200 lines, but this example has more than 400 lines.

I think this is due to the following reasons:
- it includes optional scenarios which can be executed if users edit the file, and
- it defines many variables to extract experimental settings from the execution of the experiment.

If we focus on the main scenario and directly input the variable, I think we can halve the code length. I created a prototype, so please take a look at it.
https://github.com/toshihikoyanase/optuna/blob/simplified-lgbm-w-cust-cv-fun/examples/pruning/lightgbm_with_custom_cv_function.py
